File: leetcode_21.ts
//21. Merge Two Sorted Lists

class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

let l1 = new ListNode(1, new ListNode(2, new ListNode(4)));
let l2 = new ListNode(1, new ListNode(3, new ListNode(4)));

function mergeTwoLists(
  list1: ListNode | null,
  list2: ListNode | null
): ListNode | null {
  return mergeRecursion(list1, list2);
}

function mergeRecursion(
  list1: ListNode | null,
  list2: ListNode | null
): ListNode | null {
  if (!list1 && !list2) return null;
  if ((list1?.val || 0) < (list2?.val || 0))
    return new ListNode(list1?.val, mergeRecursion(list1?.next, list2));
  else return new ListNode(list2?.val, mergeRecursion(list1, list2?.next));
}
function toNumberFormatter(list: ListNode | null): string {
  if (list === null) return "";
  else return toNumberFormatter(list.next) + list?.val;
}
console.log(toNumberFormatter(mergeTwoLists(l1, l2)));

File: leetcode_21.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func main() {
	fmt.Println(mergeTwoLists(&ListNode{1, nil}, &ListNode{1, nil}).Next.Val)
}

func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
	pre_finalListNode := ListNode{}
	ListSearcher(&pre_finalListNode, list1, list2)
	finalListNode := pre_finalListNode.Next
	return finalListNode
}

func ListSearcher(finalListNode, l1, l2 *ListNode) (*ListNode, *ListNode, *ListNode) {
	if l1 != nil && l2 != nil {
		if l1.Val > l2.Val {
			finalListNode.Next = &ListNode{Val: l2.Val, Next: nil}
			return ListSearcher(finalListNode.Next, l1, l2.Next)
		} else {
			finalListNode.Next = &ListNode{Val: l1.Val, Next: nil}
			return ListSearcher(finalListNode.Next, l1.Next, l2)
		}
	} else {
		if l1 != nil {
			finalListNode.Next = &ListNode{Val: l1.Val, Next: nil}
			return ListSearcher(finalListNode.Next, l1.Next, l2)
		} else if l2 != nil {
			finalListNode.Next = &ListNode{Val: l2.Val, Next: nil}
			return ListSearcher(finalListNode.Next, l1, l2.Next)
		} else {
			return finalListNode, nil, nil
		}
	}
}
